# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv").option("header", "true").option("inferSchema","true").load("data/housing.csv")
housing.show()

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

housing.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

housing.count()

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.take(5)

Array([-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY], [-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY], [-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY], [-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY], [-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY])

Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

housing.where("population > 10000").show()

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -121.92|   37.53|               7.0|    28258.0|        3864.0|   12203.0|    3701.0|       8.4045|          451100.0|      <1H OCEAN|
|  -117.78|   34.03|               8.0|    32054.0|        5290.0|   15507.0|    5050.0|       6.0191|          253900.0|      <1H OCEAN|
|  -117.87|   34.04|               7.0|    27700.0|        4179.0|   15037.0|    4072.0|       6.6288|          339700.0|      <1H OCEAN|
|  -117.88|   33.96|              16.0|    19059.0|        3079.0|   10988.0|    3061.0|       5.5469|          265200.0|      <1H OCEAN|
|  -118.78|   34.16|              

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

housing.select("housing_median_age", "total_rooms", "median_house_value", "population").describe().show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

housing.agg(min("total_rooms"), max("housing_median_age"), avg("median_house_value")).show()

+----------------+-----------------------+-----------------------+
|min(total_rooms)|max(housing_median_age)|avg(median_house_value)|
+----------------+-----------------------+-----------------------+
|             2.0|                   52.0|     206855.81690891474|
+----------------+-----------------------+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions.{desc, asc}

val test = housing.groupBy(housing.col("ocean_proximity")).count().orderBy(desc("count")).show()



+---------------+-----+
|ocean_proximity|count|
+---------------+-----+
|      <1H OCEAN| 9136|
|         INLAND| 6551|
|     NEAR OCEAN| 2658|
|       NEAR BAY| 2290|
|         ISLAND|    5|
+---------------+-----+



test: Unit = ()


Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy(housing.col("ocean_proximity")).avg("median_house_value").withColumnRenamed("avg(median_house_value)", "avg_value").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("SELECT ocean_proximity, AVG(median_house_value) AS avg_value FROM DF GROUP BY ocean_proximity").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.VectorAssembler

val va = new VectorAssembler().setInputCols(Array("housing_median_age", "total_rooms", "median_house_value", "population"))
.setOutputCol("attributeVector")

val housingAttrs = va.transform(housing)

housingAttrs.select("attributeVector").show(5, false)

housingAttrs.show()

+-----------------------------+
|attributeVector              |
+-----------------------------+
|[41.0,880.0,452600.0,322.0]  |
|[21.0,7099.0,358500.0,2401.0]|
|[52.0,1467.0,352100.0,496.0] |
|[52.0,1274.0,341300.0,558.0] |
|[52.0,1627.0,342200.0,565.0] |
+-----------------------------+
only showing top 5 rows

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|     attributeVector|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|

va = vecAssembler_15b595461894
housingAttrs = [longitude: double, latitude: double ... 9 more fields]


[longitude: double, latitude: double ... 9 more fields]

In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs, "attributeVector").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [13]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._


val housingCol1 = housing.withColumn("rooms_per_household", housing.col("total_rooms").divide(housing.col("households")))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", housing.col("total_bedrooms").divide(housing.col("total_rooms")))
val housingExtra = housingCol2.withColumn("population_per_household", housing.col("population").divide(housing.col("households")))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingExtra.withColumnRenamed("median_house_value", "label")
renamedHousing.printSchema

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- label: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)
 |-- rooms_per_household: double (nullable = true)
 |-- bedrooms_per_room: double (nullable = true)
 |-- population_per_household: double (nullable = true)



renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
// TODO: Replace <FILL IN> with appropriate code

for (c <- colNum) {
    println(c, renamedHousing.filter(renamedHousing.col(c).isNull).count)
}

(longitude,0)
(latitude,0)
(housing_median_age,0)
(total_rooms,0)
(total_bedrooms,207)
(population,0)
(households,0)
(median_income,0)
(rooms_per_household,0)
(bedrooms_per_room,207)
(population_per_household,0)


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median").setInputCols(Array("total_bedrooms", "bedrooms_per_room"))
.setOutputCols(Array("total_bedrooms", "bedrooms_per_room"))
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

for (c <- colNum) {
    println(c, imputedHousing.filter(imputedHousing.col(c).isNull).count)
}

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

(longitude,0)
(latitude,0)
(housing_median_age,0)
(total_rooms,0)
(total_bedrooms,0)
(population,0)
(households,0)
(median_income,0)
(rooms_per_household,0)
(bedrooms_per_room,0)
(population_per_household,0)
+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_79d06d68affe
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("features_to_scale")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setInputCol("features_to_scale").setOutputCol("features_scaled").setWithStd(true).setWithMean(true)
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5, false)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|label   |ocean_proximity|rooms_per_household|bedrooms_per_room  |population_per_household|features_to_scale                                                                                              |features_scaled                                                                                                                                                           

va = vecAssembler_643cf116d313
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_f26f8fc5b158
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [19]:
// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select("ocean_proximity").distinct.show
renamedHousing.select("ocean_proximity").distinct.count

+---------------+
|ocean_proximity|
+---------------+
|         ISLAND|
|     NEAR OCEAN|
|       NEAR BAY|
|      <1H OCEAN|
|         INLAND|
+---------------+



5

### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol("ocean_proximity")
  .setOutputCol("ocean_prox_index")

val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(25)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+----------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_prox_index|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+----------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|             3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0.15579659106

indexer = strIdx_1bb4eeddfa01
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

In [21]:
idxHousing.select("ocean_prox_index").distinct.show
idxHousing.select("ocean_prox_index").distinct.count

+----------------+
|ocean_prox_index|
+----------------+
|             0.0|
|             1.0|
|             4.0|
|             3.0|
|             2.0|
+----------------+



5

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [22]:
indexer.fit(renamedHousing).labels

Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [23]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator().setInputCols(Array("ocean_prox_index")).setOutputCols(Array("ocean_prox_onehot"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+----------------+-----------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_prox_index|ocean_prox_onehot|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+----------------+-----------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|             3.0|    (4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|

encoder = oneHotEncoder_10561b94e08e
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [24]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

//val numPipeline = 
//val catPipeline = indexer, encoder
val pipeline = new Pipeline().setStages(Array(imputer, va, scaler, indexer, encoder))
print(pipeline.getStages)
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

[Lorg.apache.spark.ml.PipelineStage;@67dd6b6e+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+----------------+-----------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|   features_to_scale|     features_scaled|ocean_prox_index|ocean_prox_onehot|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+----------------+-----------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.98412698412698

pipeline = pipeline_b1eea6968902
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

In [25]:
val newtest = newHousing.columns.filter(_ !="label").filter(_ != "ocean_proximity")
print(newtest)

[Ljava.lang.String;@7a1b45dd

newtest = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household, features_to_scale, features_scaled, ocean_prox_index, ocean_prox_onehot)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household, features_to_scale, features_scaled, ocean_prox_index, ocean_prox_onehot)

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [26]:
// TODO: Replace <FILL IN> with appropriate code

val va2 = new VectorAssembler()
    .setInputCols(newtest)
    .setOutputCol("features")
val dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-122.23,37.88,41...|452600.0|
|[-122.22,37.86,21...|358500.0|
|[-122.24,37.85,52...|352100.0|
|[-122.25,37.85,52...|341300.0|
|[-122.25,37.85,52...|342200.0|
+--------------------+--------+
only showing top 5 rows



va2 = vecAssembler_b1010da8f697
dataset = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [27]:
// TODO: Replace <FILL IN> with appropriate code
val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8, 0.2))


trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression()//.setSolver("normal").setFitIntercept(true)
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary


println(s"Coefficients: ${lrModel.coefficients}, Intercept: ${lrModel.intercept}")
println(s"RMSE: ${trainingSummary.rootMeanSquaredError}")





Coefficients: [-9179.83577881314,-8619.05328698532,362.1765044549276,0.5467307057120888,3.9231709907330146,-13.062131443237394,35.85102480676383,13711.90172423304,932.1778310780405,87090.44422642566,18.6003426420497,-9179.83577881314,-8619.05328698532,362.1765044549276,0.5467307057120888,3.9231709907330146,-13.062131443237394,35.85102480676383,13711.90172423304,932.1778310780405,87090.44422642566,18.6003426420497,-18392.091732521567,-18409.887569366696,4558.193277769645,1192.7560645947187,1645.3460462871426,-14792.369095673605,13706.913442313655,26050.168703767198,2306.3693638207865,5025.086261465561,193.1840809338243,49931.29227231985,21478.81500896521,-63678.40218622873,-74443.59432435027,-133138.87003783198], Intercept: -1564544.6551993566
RMSE: 67872.16949129598


lr = linReg_7c18471f71f7
lrModel = linReg_7c18471f71f7
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@72f2b59f


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@72f2b59f

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)


+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 182437.1599190554|111400.0|[-124.26,40.58,52...|
|183761.93269630033| 76100.0|[-124.25,40.28,32...|
|122716.59867930226| 68400.0|[-124.21,41.77,17...|
|162205.67692649667| 70200.0|[-124.18,40.78,37...|
|142433.57702170382| 81300.0|[-124.17,40.77,30...|
+------------------+--------+--------------------+
only showing top 5 rows



predictions = [features: vector, label: double ... 1 more field]


[features: vector, label: double ... 1 more field]

In [30]:
// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Root Mean Squared Error (RMSE) on test data = 67986.21580265336


evaluator = regEval_8b149025e068
rmse = 67986.21580265336


67986.21580265336

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setLabelCol("label")
  .setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)



+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|131099.27536231885|111400.0|[-124.26,40.58,52...|
|131099.27536231885| 76100.0|[-124.25,40.28,32...|
|173655.36480411046| 68400.0|[-124.21,41.77,17...|
|144172.18298446995| 70200.0|[-124.18,40.78,37...|
|144172.18298446995| 81300.0|[-124.17,40.77,30...|
+------------------+--------+--------------------+
only showing top 5 rows



dt = dtr_e51eb303235e
dtModel = DecisionTreeRegressionModel (uid=dtr_e51eb303235e) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]


[features: vector, label: double ... 1 more field]

In [32]:
// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Root Mean Squared Error (RMSE) on test data = 67986.21580265336


evaluator = regEval_285d099fce63
rmse = 67986.21580265336


67986.21580265336

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [33]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setLabelCol("label")
  .setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)


+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 151817.8955883835|111400.0|[-124.26,40.58,52...|
|160389.37036499396| 76100.0|[-124.25,40.28,32...|
|155064.43085837772| 68400.0|[-124.21,41.77,17...|
|150369.42339362684| 70200.0|[-124.18,40.78,37...|
|145066.98153420317| 81300.0|[-124.17,40.77,30...|
+------------------+--------+--------------------+
only showing top 5 rows



rf = rfr_5f48b5c31aeb
rfModel = RandomForestRegressionModel (uid=rfr_5f48b5c31aeb) with 20 trees
predictions = [features: vector, label: double ... 1 more field]


[features: vector, label: double ... 1 more field]

In [34]:
// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Root Mean Squared Error (RMSE) on test data = 67384.8874309743


evaluator = regEval_c4e5c4215304
rmse = 67384.8874309743


67384.8874309743

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [35]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setLabelCol("label")
  .setFeaturesCol("features")
  
// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)


+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 79341.57054607717|111400.0|[-124.26,40.58,52...|
|107832.26614868997| 76100.0|[-124.25,40.28,32...|
|118476.86342071832| 68400.0|[-124.21,41.77,17...|
| 78931.77589044173| 70200.0|[-124.18,40.78,37...|
| 84782.23716231633| 81300.0|[-124.17,40.77,30...|
+------------------+--------+--------------------+
only showing top 5 rows



gb = gbtr_1e860a3f8cee
gbModel = GBTRegressionModel (uid=gbtr_1e860a3f8cee) with 20 trees
predictions = [features: vector, label: double ... 1 more field]


[features: vector, label: double ... 1 more field]

In [36]:

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Root Mean Squared Error (RMSE) on test data = 63396.716666605935


evaluator = regEval_b41fb4ff057a
rmse = 63396.716666605935


63396.716666605935

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [37]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder().addGrid(rf.numTrees , Array(1, 5, 10))
  .addGrid(rf.maxDepth ,Array(5, 10, 15))
  .build()

val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val cv = new CrossValidator().setEstimator(rf)
  .setEvaluator(evaluator)
  .setEstimatorParamMaps(paramGrid)
  .setNumFolds(3)  // Use 3+ in practice
val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")






+----------------+--------+--------------------+
|      prediction|   label|            features|
+----------------+--------+--------------------+
|         93666.0|111400.0|[-124.26,40.58,52...|
|        115040.0| 76100.0|[-124.25,40.28,32...|
|        116552.0| 68400.0|[-124.21,41.77,17...|
|         81615.0| 70200.0|[-124.18,40.78,37...|
|80606.0476190476| 81300.0|[-124.17,40.77,30...|
+----------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 52384.557143718375


paramGrid = 


Array({
	rfr_5f48b5c31aeb-maxDepth: 5,
	rfr_5f48b5c31aeb-numTrees: 1
}, {
	rfr_5f48b5c31aeb-maxDepth: 5,
	rfr_5f48b5c31aeb-numTrees: 5
}, {
	rfr_5f48b5c31aeb-maxDepth: 5,
	rfr_5f48b5c31aeb-numTrees: 10
}, {
	rfr_5f48b5c31aeb-maxDepth: 10,
	rfr_5f48b5c31aeb-numTrees: 1
}, {
	rfr_5f48b5c31aeb-maxDepth: 10,
	rfr_5f48b5c31aeb-numTrees: 5
}, {
	rfr_5f48b5c31aeb-maxDepth: 10,
	rfr_5f48b5c31aeb-numTrees: 10
}, {
	rfr_5f48b5c31aeb-maxDepth: 15,
	rfr_5f48b5c31aeb-numTrees: 1
}, {
	rfr_5f48b5c31aeb-maxDepth: 15,
	rfr_5f48b5c31aeb-numTrees: 5
}, {
	rfr_5f48b5c31aeb-maxDepth: 15,
	rfr_5f48...


---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [40]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "The input column 1")
    final val inputCol2 = new Param[String](this, "inputCol2", "The input column 2")
    final val outputCol = new Param[String](this, "outputCol", "The output column")
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [41]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.functions.{col, udf}

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("MyTransformer"))
    
    def setInputCol1(value: String) = set(inputCol1, value)
    
    def setInputCol2(value: String) = set(inputCol2, value)
    
    def setOutputCol(value: String) = set(outputCol, value)

    override def copy(extra: ParamMap): MyTransformer = {
        defaultCopy(extra)
    }
    
    override def transformSchema(schema: StructType) = {
        // Check that the input type is a double
        val idx1 = schema.fieldIndex($(inputCol1))
        val idx2 = schema.fieldIndex($(inputCol2))

        val field1 = schema.fields(idx1)
        if (field1.dataType != DoubleType) {
          throw new Exception(s"Input type ${field1.dataType} did not match input type DoubleType")
        }
        val field2 = schema.fields(idx2)
        if (field2.dataType != DoubleType) {
          throw new Exception(s"Input type ${field2.dataType} did not match input type DoubleType")
        }

          // Add the return field
        schema.add(StructField($(outputCol), DoubleType, false))
    }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        dataset.select(col("*"),
            dataset($(inputCol1)).divide(dataset($(inputCol2))).as($(outputCol)))
    }
}


defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [42]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = MyTransformer_171e78816d10


myDataset: Unit = ()


MyTransformer_171e78816d10

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [43]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [45]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [47]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [48]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [49]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-3.58489318109883...| 94600.0|[-124.35,40.54,52...|
| -5.4058946585443E76| 85800.0|[-124.3,41.8,19.0...|
|-5.35865637789578...|103600.0|[-124.3,41.84,17....|
|-4.81997549848366...| 79000.0|[-124.27,40.69,36...|
|-5.28341710216546...|106700.0|[-124.23,40.54,52...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 6.93774523645173E76


lr = linReg_c5cb2a488557
model = linReg_c5cb2a488557
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_371ff8a4dc48
rmse = 6.93774523645173E76


6.93774523645173E76

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

In [51]:
// 1. Load the data
val creds = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load("data/ccdefault.csv")

//creds.show
creds.printSchema

root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- EDUCATION: integer (nullable = true)
 |-- MARRIAGE: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_0: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- DEFAULT: integer (nullable = tru

creds = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

In [53]:
// 2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).

// Summary of attributes: limitBal, age

creds.select("LIMIT_BAL", "AGE").describe().show

+-------+------------------+-----------------+
|summary|         LIMIT_BAL|              AGE|
+-------+------------------+-----------------+
|  count|             30000|            30000|
|   mean|167484.32266666667|          35.4855|
| stddev|129747.66156720246|9.217904068090155|
|    min|             10000|               21|
|    max|           1000000|               79|
+-------+------------------+-----------------+



In [65]:
// 15 Largest credit defaults
creds.select("LIMIT_BAL", "DEFAULT").where(creds.col("DEFAULT") === 1).orderBy(desc("LIMIT_BAL")).show(15)

+---------+-------+
|LIMIT_BAL|DEFAULT|
+---------+-------+
|   740000|      1|
|   720000|      1|
|   710000|      1|
|   680000|      1|
|   630000|      1|
|   620000|      1|
|   600000|      1|
|   600000|      1|
|   600000|      1|
|   590000|      1|
|   580000|      1|
|   560000|      1|
|   550000|      1|
|   550000|      1|
|   550000|      1|
+---------+-------+
only showing top 15 rows



In [74]:
// Most commonly defaulted ages
creds.select("AGE", "DEFAULT").groupBy(creds.col("AGE")).count().orderBy(desc("count")).show

+---+-----+
|AGE|count|
+---+-----+
| 29| 1605|
| 27| 1477|
| 28| 1409|
| 30| 1395|
| 26| 1256|
| 31| 1217|
| 25| 1186|
| 34| 1162|
| 32| 1158|
| 33| 1146|
| 24| 1127|
| 35| 1113|
| 36| 1108|
| 37| 1041|
| 39|  954|
| 38|  944|
| 23|  931|
| 40|  870|
| 41|  824|
| 42|  794|
+---+-----+
only showing top 20 rows



lastException: Throwable = null


In [78]:
// null-check
for (c <- creds.columns) {
    println(c, creds.filter(creds.col(c).isNull).count)
}

(ID,0)
(LIMIT_BAL,0)
(SEX,0)
(EDUCATION,0)
(MARRIAGE,0)
(AGE,0)
(PAY_0,0)
(PAY_2,0)
(PAY_3,0)
(PAY_4,0)
(PAY_5,0)
(PAY_6,0)
(BILL_AMT1,0)
(BILL_AMT2,0)
(BILL_AMT3,0)
(BILL_AMT4,0)
(BILL_AMT5,0)
(BILL_AMT6,0)
(PAY_AMT1,0)
(PAY_AMT2,0)
(PAY_AMT3,0)
(PAY_AMT4,0)
(PAY_AMT5,0)
(PAY_AMT6,0)
(DEFAULT,0)


In [86]:
// wrangle data

val colNumCreds = creds.columns.filter(_ != "DEFAULT")

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va2 = new VectorAssembler().setInputCols(colNumCreds).setOutputCol("features_to_scale")

val featuredCreds = va.transform(creds)

val scaler = new StandardScaler().setInputCol("features_to_scale").setOutputCol("features_scaled").setWithStd(true).setWithMean(true)
val scaledCreds = scaler.fit(featuredCreds).transform(featuredCreds)

scaledCreds.show(5, false)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|ID |LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|P

colNumCreds = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)
va2 = vecAssembler_95ebaf803d84
featuredCreds = [ID: int, LIMIT_BAL: int ... 24 more fields]
scaler = stdScal_8734d53d45d1
scaledCreds = [ID: int, LIMIT_BAL: int ... 25 more fields]


[ID: int, LIMIT_BAL: int ... 25 more fields]

In [87]:
val featuresColsCreds = scaledCreds.columns.filter(_ != "DEFAULT")

import org.apache.spark.ml.feature.VectorAssembler

val va3 = new VectorAssembler().setInputCols(featuresColsCreds).setOutputCol("features")



featuresColsCreds = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, features_to_scale, features_scaled)
va3 = vecAssembler_9e659d07f42d


vecAssembler_9e659d07f42d

In [100]:
val dataset2 = va3.transform(scaledCreds).select("features", "DEFAULT")
val datasetRenamed = dataset2.withColumnRenamed("DEFAULT", "label")
datasetRenamed.show

+--------------------+-----+
|            features|label|
+--------------------+-----+
|[1.0,20000.0,2.0,...|    1|
|[2.0,120000.0,2.0...|    1|
|[3.0,90000.0,2.0,...|    0|
|[4.0,50000.0,2.0,...|    0|
|[5.0,50000.0,1.0,...|    0|
|[6.0,50000.0,1.0,...|    0|
|[7.0,500000.0,1.0...|    0|
|[8.0,100000.0,2.0...|    0|
|[9.0,140000.0,2.0...|    0|
|[10.0,20000.0,1.0...|    0|
|[11.0,200000.0,2....|    0|
|[12.0,260000.0,2....|    0|
|[13.0,630000.0,2....|    0|
|[14.0,70000.0,1.0...|    1|
|[15.0,250000.0,1....|    0|
|[16.0,50000.0,2.0...|    0|
|[17.0,20000.0,1.0...|    1|
|[18.0,320000.0,1....|    0|
|[19.0,360000.0,2....|    0|
|[20.0,180000.0,2....|    0|
+--------------------+-----+
only showing top 20 rows



dataset2 = [features: vector, DEFAULT: int]
datasetRenamed = [features: vector, label: int]


lastException: Throwable = null


[features: vector, label: int]

In [101]:
// Split the data
val Array(trainSetCreds, testSetCreds) = datasetRenamed.randomSplit(Array(0.8, 0.2))

trainSetCreds = [features: vector, label: int]
testSetCreds = [features: vector, label: int]


[features: vector, label: int]

In [124]:
// logistic regression
import org.apache.spark.ml.classification.LogisticRegression

val lr = new LogisticRegression()
  .setMaxIter(10)

// Fit the model
val lrModel = lr.fit(trainSetCreds)

// Print the coefficients and intercept for logistic regression
println(s"Coefficients: ${lrModel.coefficients} Intercept: ${lrModel.intercept}")

// We can also use the multinomial family for binary classification
val mlr = new LogisticRegression()
  .setMaxIter(10)
  .setFamily("multinomial")

val mlrModel = mlr.fit(trainSetCreds)

// Print the coefficients and intercepts for logistic regression with multinomial family
println(s"Multinomial coefficients: ${mlrModel.coefficientMatrix}")
println(s"Multinomial intercepts: ${mlrModel.interceptVector}")



Coefficients: [2.8602411997530166E-7,-2.7257078443749193E-7,-0.023979670044358588,-0.022831737180365574,-0.03353239651027701,0.004129620971474542,0.1991393327157488,0.02106138041629552,0.01957695206009359,0.01715360859883148,0.009593445318697455,0.004262998048346987,-4.6833228181426344E-7,-1.4666948649299066E-7,-4.137394791277771E-8,6.985492517377559E-8,2.021442868958358E-7,2.678780356887319E-7,-3.6311492859733626E-6,-2.0843334381368446E-6,-1.1255854573867622E-6,-1.0027342322793153E-6,-9.921235886076663E-7,-7.766737903321108E-7,2.8602411997530166E-7,-2.7257078443749193E-7,-0.023979670044358588,-0.022831737180365574,-0.03353239651027701,0.004129620971474542,0.1991393327157488,0.02106138041629552,0.01957695206009359,0.01715360859883148,0.009593445318697455,0.004262998048346987,-4.6833228181426344E-7,-1.4666948649299066E-7,-4.137394791277771E-8,6.985492517377559E-8,2.021442868958358E-7,2.678780356887319E-7,-3.6311492859733626E-6,-2.0843334381368446E-6,-1.1255854573867622E-6,-1.00273423227

lr = logreg_6b44c5a46751
lrModel = logreg_6b44c5a46751
mlr = logreg_56b2446098fe
mlrModel = logreg_56b2446098fe


logreg_56b2446098fe

In [125]:
// Extract the summary from the returned LogisticRegressionModel instance trained in the earlier
// example
val trainingSummary = lrModel.binarySummary

// Obtain the objective per iteration.
val objectiveHistory = trainingSummary.objectiveHistory
println("objectiveHistory:")
objectiveHistory.foreach(loss => println(loss))

// Obtain the receiver-operating characteristic as a dataframe and areaUnderROC.
val roc = trainingSummary.roc
roc.show()
println(s"areaUnderROC: ${trainingSummary.areaUnderROC}")

// Set the model threshold to maximize F-Measure
val fMeasure = trainingSummary.fMeasureByThreshold
val maxFMeasure = fMeasure.select(max("F-Measure")).head().getDouble(0)
val bestThreshold = fMeasure.where($"F-Measure" === maxFMeasure)
  .select("threshold").head().getDouble(0)
lrModel.setThreshold(bestThreshold)

println(s"Accuracy: ${trainingSummary.accuracy}")


objectiveHistory:
0.5304054706330218
+---+---+
|FPR|TPR|
+---+---+
|0.0|0.0|
|1.0|1.0|
|1.0|1.0|
+---+---+

areaUnderROC: 0.5
Accuracy: 0.7772188734150904


trainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@2217f009
objectiveHistory = Array(0.5304054706330218)
roc = [FPR: double, TPR: double]
fMeasure = [threshold: double, F-Measure: double]
maxFMeasure = 0.3643843067926838
bestThreshold = 0.2227811265849096


0.2227811265849096

In [126]:
// Extract the summary from the returned LogisticRegressionModel instance trained in the earlier
// example
val trainingSummary = mlrModel.binarySummary

// Obtain the objective per iteration.
val objectiveHistory = trainingSummary.objectiveHistory
println("objectiveHistory:")
objectiveHistory.foreach(loss => println(loss))

// Obtain the receiver-operating characteristic as a dataframe and areaUnderROC.
val roc = trainingSummary.roc
roc.show()
println(s"mlr areaUnderROC: ${trainingSummary.areaUnderROC}")

// Set the model threshold to maximize F-Measure
val fMeasure = trainingSummary.fMeasureByThreshold
val maxFMeasure = fMeasure.select(max("F-Measure")).head().getDouble(0)
val bestThreshold = fMeasure.where($"F-Measure" === maxFMeasure)
  .select("threshold").head().getDouble(0)
lrModel.setThreshold(bestThreshold)

println(s"MLR Accuracy: ${trainingSummary.accuracy}")


objectiveHistory:
0.5304054721664393
0.485178948280689
0.481425674332408
0.47673519110011653
0.4727164756710272
0.46823669913201565
0.4661362646907135
0.46547704795547534
0.465278906551422
0.46510509529971733
0.4649706132355405
+--------------------+-------------------+
|                 FPR|                TPR|
+--------------------+-------------------+
|                 0.0|                0.0|
|0.003369704749679...|0.03302855010263109|
|0.007113821138211382|0.06475088635939541|
| 0.00978818998716303|0.10020526217577906|
|0.013746255883611468|0.13118119052061952|
|0.017276422764227643|0.16364993468930772|
| 0.02064612751390672| 0.1966784847919388|
| 0.02471116816431322|0.22728120918081732|
| 0.02920410783055199| 0.2563911177458481|
| 0.03449935815147625| 0.2827019966411644|
| 0.03925973470261018|0.31087889531629037|
| 0.04444801026957638| 0.3375629781675686|
| 0.05059905862216517| 0.3608882254151894|
| 0.05733846812152332| 0.3821608509050196|
|  0.0641313649978605| 0.4032468744168688

trainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@7785ed7e
objectiveHistory = Array(0.5304054721664393, 0.485178948280689, 0.481425674332408, 0.47673519110011653, 0.4727164756710272, 0.46823669913201565, 0.4661362646907135, 0.46547704795547534, 0.465278906551422, 0.46510509529971733, 0.4649706132355405)
roc = [FPR: double, TPR: double]
fMeasure = [threshold: double, F-Measure: double]
maxFMeasure = 0.5215080224431539
bestThreshold = 0.2953450366952304


0.2953450366952304

In [127]:
// make predictions on the test data LRModel
val predictions = lrModel.transform(testSetCreds)
predictions.select("prediction", "label", "features").show(5)

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
+----------+-----+--------------------+
only showing top 5 rows



predictions = [features: vector, label: int ... 3 more fields]


[features: vector, label: int ... 3 more fields]

In [128]:
// Select (prediction, true label) and compute test error.
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)
println(s"Test Error = ${(1.0 - accuracy)}")


Test Error = 0.2148023549201009


evaluator = mcEval_2afd02ff7fcb
accuracy = 0.7851976450798991


0.7851976450798991

In [129]:
// make predictions on the test data MLRmodel
val predictionsmlr = mlrModel.transform(testSetCreds)
predictionsmlr.select("prediction", "label", "features").show(5)

+----------+-----+--------------------+
|prediction|label|            features|
+----------+-----+--------------------+
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
|       0.0|    0|(72,[0,1,2,3,4,5,...|
+----------+-----+--------------------+
only showing top 5 rows



predictionsmlr = [features: vector, label: int ... 3 more fields]


[features: vector, label: int ... 3 more fields]

In [130]:
// Select (prediction, true label) and compute test error.
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictionsmlr)
println(s"Test Error = ${(1.0 - accuracy)}")


Test Error = 0.2148023549201009


evaluator = mcEval_0cd5dfa7ec1e
accuracy = 0.7851976450798991


0.7851976450798991

In [141]:
// decision tree
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val dt = new DecisionTreeClassifier()
  .setLabelCol("label")
  .setFeaturesCol("features")

// Train model. This also runs the indexers.
val dtmodel = dt.fit(trainSetCreds)

// Make predictions.
val predictions = dtmodel.transform(testSetCreds)

// Select example rows to display.
predictions.show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)
println(s"Test Error = ${(1.0 - accuracy)}")

println(s"Learned classification tree model:\n ${dtmodel.toDebugString}")


+--------------------+-----+--------------+--------------------+----------+
|            features|label| rawPrediction|         probability|prediction|
+--------------------+-----+--------------+--------------------+----------+
|(72,[0,1,2,3,4,5,...|    0|[1303.0,230.0]|[0.84996738421395...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[1303.0,230.0]|[0.84996738421395...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[1303.0,230.0]|[0.84996738421395...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[1303.0,230.0]|[0.84996738421395...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[1303.0,230.0]|[0.84996738421395...|       0.0|
+--------------------+-----+--------------+--------------------+----------+
only showing top 5 rows

Test Error = 0.1754415475189235
Learned classification tree model:
 DecisionTreeClassificationModel (uid=dtc_5e169b20e9ae) of depth 5 with 61 nodes
  If (feature 6 <= 1.5)
   If (feature 7 <= 1.5)
    If (feature 20 <= 1.5)
     If (feature 12 <= 0.5)
      If (feature 5 <= 55.5)
       Pre

dt = dtc_5e169b20e9ae
dtmodel = DecisionTreeClassificationModel (uid=dtc_5e169b20e9ae) of depth 5 with 61 nodes
predictions = [features: vector, label: int ... 3 more fields]
evaluator = mcEval_3300f2e724b7
accuracy = 0.8245584524810765


0.8245584524810765

In [142]:
// random forest
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}

// Train a RandomForest model.
val rf = new RandomForestClassifier()
  .setLabelCol("label")
  .setFeaturesCol("features")
  .setNumTrees(10)

// Train model. This also runs the indexers.
val rfmodel = rf.fit(trainSetCreds)

// Make predictions.
val predictions = rfmodel.transform(testSetCreds)

// Select example rows to display.
predictions.show(15)

// Select (prediction, true label) and compute test error.
val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("label")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(predictions)
println(s"Test Error = ${(1.0 - accuracy)}")

println(s"Learned classification forest model:\n ${rfmodel.toDebugString}")


+--------------------+-----+--------------------+--------------------+----------+
|            features|label|       rawPrediction|         probability|prediction|
+--------------------+-----+--------------------+--------------------+----------+
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.09104967577805...|[0.80910496757780...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,5,...|    0|[8.03139736344747...|[0.80313973634474...|       0.0|
|(72,[0,1,2,3,4,

rf = rfc_220393a36462
rfmodel = RandomForestClassificationModel (uid=rfc_220393a36462) with 10 trees
predictions = [features: vector, label: int ... 3 more fields]
evaluator = mcEval_5bcf154ad939
accuracy = 0.8205214465937762


0.8205214465937762